runtime/src/bank/fee_distribution.rs

```diff
@@ -0,0 +1,893 @@
+        // If two items are still equal, their relative order does not matter since
+        // both refer to the same validator.
+        validator_stakes.sort_unstable_by(|(pubkey1, staked1), (pubkey2, staked2)| {
+            (staked1, pubkey1).cmp(&(staked2, pubkey2)).reverse()
+        });
+
+        let enforce_fix = self.no_overflow_rent_distribution_enabled();
+
+        let mut rent_distributed_in_initial_round = 0;
+        let validator_rent_shares = validator_stakes
+            .into_iter()
+            .map(|(pubkey, staked)| {
+                let rent_share = if !enforce_fix {
+                    (((staked * rent_to_be_distributed) as f64) / (total_staked as f64)) as u64
+                } else {
+                    (((staked as u128) * (rent_to_be_distributed as u128)) / (total_staked as u128))
+                        .try_into()
+                        .unwrap()
+                };
+                rent_distributed_in_initial_round += rent_share;
+                (pubkey, rent_share)
+            })
+            .collect::<Vec<(Pubkey, u64)>>();
+
+        // Leftover lamports after fraction calculation, will be paid to validators starting from highest stake
+        // holder
+        let mut leftover_lamports = rent_to_be_distributed - rent_distributed_in_initial_round;
+
+        let mut rewards = vec![];
+        validator_rent_shares
+            .into_iter()
+            .for_each(|(pubkey, rent_share)| {
+                let rent_to_be_paid = if leftover_lamports > 0 {
+                    leftover_lamports -= 1;
+                    rent_share + 1
+                } else {
+                    rent_share
+                };
+                if !enforce_fix || rent_to_be_paid > 0 {
+                    let check_account_owner = self.validate_fee_collector_account();
+                    let check_rent_paying = self.prevent_rent_paying_rent_recipients();
+                    match self.deposit_fees(
+                        &pubkey,
+                        rent_to_be_paid,
+                        DepositFeeOptions {
+                            check_account_owner,
+                            check_rent_paying,
+                        },
+                    ) {
+                        Ok(post_balance) => {
+                            rewards.push((
+                                pubkey,
+                                RewardInfo {
+                                    reward_type: RewardType::Rent,
+                                    lamports: rent_to_be_paid as i64,
+                                    post_balance,
+                                    commission: None,
+                                },
+                            ));
+                        }
+                        Err(err) => {
+                            // overflow adding lamports or resulting account is invalid
+                            self.capitalization.fetch_sub(rent_to_be_paid, Relaxed);
```

Comment:
Instead of burning the fees here, do we consider another option - we distribute these rents to other staked nodes? yet another thought ...

Reply:
Yeah it's a good thought but I think it's not worth the effort and complexity. Transaction fees shouldn't be distributed to other staked nodes because they only go to the block producer. Rent fees could be re-distributed but they will be phased out over time once there are basically no more rent paying accounts left on the cluster.